jgrapht-core/src/main/java/org/jgrapht/alg/matching/blossom/v5/Node.java

```diff
@@ -0,0 +1,531 @@
+     *
+     * @param child      the new child of this node
+     * @param parentEdge the edge between this node and {@code child}
+     * @param grow       true if {@code child} is being grown
+     */
+    public void addChild(Node child, Edge parentEdge, boolean grow) {
+        child.parentEdge = parentEdge;
+        child.tree = tree;
+        child.treeSiblingNext = firstTreeChild;
+        if (grow) {
+            // if child is being grown => we have to overwrite all its tree structure data
+            child.firstTreeChild = null;
+        }
+        if (firstTreeChild == null) {
+            child.treeSiblingPrev = child;
+        } else {
+            child.treeSiblingPrev = firstTreeChild.treeSiblingPrev;
+            firstTreeChild.treeSiblingPrev = child;
+        }
+        firstTreeChild = child;
+    }
+
+    /**
+     * Helper method, returns a node this node is matched to.
+     *
+     * @return a node this node is matched to.
+     */
+    public Node getOppositeMatched() {
+        return matched.getOpposite(this);
+    }
+
+    /**
+     * If this node is a tree root then this method removes this nodes from the tree root doubly linked list.
+     * Otherwise, removes this vertex from the doubly linked list of tree children and updates
+     * parent.firstTreeChild accordingly.
+     */
+    public void removeFromChildList() {
+        if (isTreeRoot) {
+            treeSiblingPrev.treeSiblingNext = treeSiblingNext;
+            if (treeSiblingNext != null) {
+                treeSiblingNext.treeSiblingPrev = treeSiblingPrev;
+            }
+        } else {
+            if (treeSiblingPrev.treeSiblingNext == null) {
+                // this vertex is the first child => we have to update parent.firstTreeChild
+                parentEdge.getOpposite(this).firstTreeChild = treeSiblingNext;
+            } else {
+                // this vertex isn't the first child
+                treeSiblingPrev.treeSiblingNext = treeSiblingNext;
+            }
+            if (treeSiblingNext == null) {
+                // this vertex is the last child => we have to set treeSiblingPrev of the firstChild
+                if (parentEdge.getOpposite(this).firstTreeChild != null) {
+                    parentEdge.getOpposite(this).firstTreeChild.treeSiblingPrev = treeSiblingPrev;
+                }
+            } else {
+                // this vertex isn't the last child
+                treeSiblingNext.treeSiblingPrev = treeSiblingPrev;
+            }
+        }
+    }
+
+    /**
+     * Appends the child list of this node to the beginning of the child list of the {@code blossom}.
+     *
+     * @param blossom the node to which the children of the current node are moved
+     */
+    public void moveChildrenTo(Node blossom) {
+        if (firstTreeChild != null) {
+            if (blossom.firstTreeChild == null) {
+                blossom.firstTreeChild = firstTreeChild;
+            } else {
+                Node first = firstTreeChild;
```

Comment:
You are holding on to this because it could be the same as below and hence overwritten? When would this happen?

That aside, I would consider changing the name (maybe `firstTreeChildRef`?) so it doesn't conflict with Class variable `first`

Reply:
This is rather a bug